Enhance separation between in-toto-verify cli tool and library

lukpueh opened the pull request:
Moves complete in_toto_verify workflow to verifylib.in_toto_verify

The user-facing in-toto-verify command line tool just takes care of command line argument parsing and calls verifylib.in_toto_verify. In case an exception occurs it is caught in the calling CLI tool, which exits gracefully.

- Adds an optional --verbose flag to in-toto-verify CLI tool that turns on logging statements in the library.
- Enhances logging statements in verifylib.
- Changes CLI argument parsing for multiple layout keys.
- Adds a copy of metadata files from in_toto demo to unittest directory for verification testing
- in_toto_verify/verifylib is now similarly structured as in_toto_record/runlib. See discussion in #6 (comment).

- Takes over https://github.com/in-toto/in-toto/pull/40

Commit:
Moves complete verify workflow to verifylib

- The user-facing in-toto-verify command line tool just takes care
  of argument parsing and calls verifylib.in_toto_verify, which
  carries out all verification steps. If an exception occurs it
  is caught in the CLI tool, which exits gracefully

- Adds an optional --verbose flag to in-toto-run CLI tool that turns on logging
  statements in the library.

- Enhances logging statements on verification.

- in_toto_run/runlib is now similarly structured as in_toto_record/runlib.

Commit:
Adds final product test files to unittest dir

Takes final product as produced by demo and adds passing and
failing supply chain metadata to unittest directory to be used
by unittests.
Also adds a double signed root layout to test verification with
multiple project owner keys.

Commit:
Removes whitespace

Commit:
Adds in-toto-verify command line tool unittests

Test the command line parser in_toto_verify.main and the cli
tools function that calls into verifylib and handles exceptions

Commit:
Updates test_verifylib and test_in_toto_verify

Removes demo_fail and demo_pass test file directories and adds
demo_files directory instead.
Demo files contains all link metadata files, functionary and
project owner keys and the final product as created in the
in-toto demo. Plus it contains a layout template (not signed,
no expiration date).

test_verifylib and test_in_toto_verify use those metadatafiles
and keys to test passing and failing supply chains.
The layout template is read and re-dumped with various different
parameters (e.g. passed date to test expiration, multiple signatures,
failing match rules, ...)

Commit:
Merges exceptions in from import statement

Commit:
Updates docstrings, fixes bug in test_verifylib

Fixes failing inspection matchrule test

Commit:
Updates docstrings in test_in_toto_verify

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -593,4 +593,84 @@ def verify_all_item_rules(items, links, target_links=None):
+    by performing the following actions:
+        1.  Load root layout as specified by 'layout_path'
+        2.  Load root layout signing key(s) (project owner public key(s)) as
+            specified by 'layout_key_paths'
```
layout_key_paths is a list? why is it plural?

lukpueh replied:
`layout_key_paths` is a list that contains paths to public keys whose private portions were used to sign the layout. A layout can be signed by multiple keys, right?

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -593,4 +593,84 @@ def verify_all_item_rules(items, links, target_links=None):
+            NOTE: in_toto_verify searches for link metadata files by their step
+            name plus the extension '.link' in the current directory.
+            E.g. a step of name 'foo' is expected to be found in the current
+            directory at 'foo.link'.
```
recall that we now use part of the keyid as a postfix (we could glob this out).

lukpueh replied:
Not yet. The keyid postfix will be in PR #63 or #64. :)  Which will also have to update this function (plus docstrings).

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -593,4 +593,84 @@ def verify_all_item_rules(items, links, target_links=None):
+            name plus the extension '.link' in the current directory.
+            E.g. a step of name 'foo' is expected to be found in the current
+            directory at 'foo.link'.
+        6.  For link metadata signatures
```
should the following steps be indented?

lukpueh replied:
No, but the sentence seems to be incomplete. Thanks for spotting that something is odd.

SantiagoTorres replied:
ping

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -593,4 +593,84 @@ def verify_all_item_rules(items, links, target_links=None):
+  <Side Effects>
+    Reads layout from disk
+    Reads project owner public key(s) from disk
+    Reads link metadata files from disk
```
Maybe I'm wrong, but is reading a side effect? (I don't know what's the standard for this).

lukpueh replied:
I'm never sure about these either. Just talked about it with @vladimir-v-diaz. He partially agrees with you. Then again, one could argue that reading modifies a file's `atime`.    Anyway, I guess it should be noted but it does not matter if it is called side effect or not, so I can move these operations to the `<Purpose>` section.

SantiagoTorres replied:
I'd advice to just nix it, but I"m not married with the idea

SantiagoTorres commented on test/test_in_toto_verify.py:
```diff
@@ -0,0 +1,157 @@
+
+  Copies the basic layout for different test scenarios:
+    - single-signed layout
+    - double-signed layout
```
I don't know if I'm missing something. What do we mean by double-signed layout?

lukpueh replied:
Maybe this is not in the specification, but the current implementation allows to sign (and verify) a layout with multiple keys (see also https://github.com/in-toto/in-toto/pull/62#discussion_r96055667)

SantiagoTorres replied:
This is fine (it's not specified yet), and I don't see it as a blocker. I think it'd be clearer if we called this (multiple-signatures-on-layout) and then specify it's 2 to convey the fact that it supports multiple signature, not just one case or the other.

SantiagoTorres commented on test/test_verifylib.py:
```diff
@@ -584,6 +588,157 @@ def test_pass_verify_all_inspection_rules(self):
+
+  def test_verify_failing_inspection_exits_non_zero(self):
+    """FIXME implement raise exception on inspection fail. """
+    pass
```
Is this missing now? is there a ticket for this?

lukpueh replied:
Correct and correct. We discussed this with the VCS AppSec group, since they would have appreciated mentioned behavior. They planned to submit a ticket but never did.

SantiagoTorres replied:
hmm, could we ticketize this then?

lukpueh replied:
https://github.com/in-toto/in-toto/issues/63

SantiagoTorres reviewed (commented):
I've just got a couple of questions, but I don't think any of these is a blocker.

Commit:
Fixes docstring bug in verifylib

Commit:
Removes side effect docstring from in_toto_verify

It can be argued that reading a file is a side effect, since it
changes some global state (e.g. atime). But maybe that's too
far fetched.
For now we nix it and mention the operation in the purpose section
of the function's docstrings.

Commit:
Changes double to multi signed layout in test docs
